Doc/tools/extensions/pyspecific.py

```diff
@@ -569,10 +570,13 @@ def process_audit_events(app, doctree, fromdocname):
         for i, (doc, label) in backlinks:
             if isinstance(label, str):
                 ref = nodes.reference("", nodes.Text("[{}]".format(i)), internal=True)
-                ref['refuri'] = "{}#{}".format(
-                    app.builder.get_relative_uri(fromdocname, doc),
-                    label,
-                )
+                try:
+                    ref['refuri'] = "{}#{}".format(
+                        app.builder.get_relative_uri(fromdocname, doc),
+                        label,
+                    )
+                except NoUri:
```

Comment:
Perhaps to avoid the import and provide useful feedback?

```suggestion
                except Exception as ex:
                    logging.getLogger("pyspecific.AuditEvent").warn(
                        "Failed to generate link: {}".format(ex))
```

Reply:
I prefer being precise and ignore `NoUri`, as Sphinx does interally. It's fully expected to get a `NoUri` here, it's not "exceptional": we're buildin a PDF file per section (library, c-api, tutorial, ...) but we're trying to link from a section to another (from library to c-api) so the links can't work. Other exceptions are *not* normal (I don't yet know them), so I prefer for them to bubble up explicitly.

Reply:
Agreed. Having both imports is the better option.